File: LeetCode/50-pow(x,n)/Solution.java
/**
 * This class contains methods for calculating the power of a number using
 * binary exponentiation.
 */
public class Solution {

  /**
   * Computes x raised to the power n using binary exponentiation.
   *
   * @param x the base
   * @param n the exponent
   * @return the result of x raised to the power n
   */
  public double binaryExp(double x, long n) {
    // Base case to stop recursive calls.
    if (n == 0) {
      return 1;
    }

    // Handle case where n < 0.
    if (n < 0) {
      return 1.0 / binaryExp(x, -n);
    }

    // Perform Binary Exponentiation.
    // If 'n' is odd, perform Binary Exponentiation on 'n - 1' and multiply result
    // with 'x'.
    if (n % 2 == 1) {
      return x * binaryExp(x * x, (n - 1) / 2);
    } else {
      // Otherwise calculate result by performing Binary Exponentiation on 'n'.
      return binaryExp(x * x, n / 2);
    }
  }

  /**
   * Computes x raised to the power n.
   *
   * @param x the base
   * @param n the exponent
   * @return the result of x raised to the power n
   */
  public double myPow(double x, int n) {
    return binaryExp(x, (long) n);
  }
}

File: LeetCode/50-pow(x,n)/solution.ts
/**
 * Computes x raised to the power n using iterative binary exponentiation.
 *
 * @param {number} x - The base number.
 * @param {number} n - The exponent.
 * @returns {number} - The result of x raised to the power n.
 */
function myPow(x: number, n: number): number {
  /**
   * Helper function for binary exponentiation.
   *
   * @param {number} x - The base number.
   * @param {number} n - The exponent.
   * @returns {number} - The result of x raised to the power n.
   */
  function binaryExp(x: number, n: number): number {
    // Base case: If n is 0, return 1
    if (n === 0) {
      return 1;
    }

    // Handle case where n < 0: Invert x and make n positive
    if (n < 0) {
      n = -n;
      x = 1.0 / x;
    }

    let result = 1; // Initialize result to 1
    while (n > 0) {
      // If n is odd, multiply the current result by x and reduce n by 1
      if (n % 2 === 1) {
        result *= x;
        n -= 1;
      }
      // Square x and reduce n by half
      x *= x;
      n = Math.floor(n / 2);
    }

    return result; // Return the final result
  }

  return binaryExp(x, n); // Call the helper function and return its result
};
